#164139774 Create new recipe

blackdevelopa opened the pull request:
**What does this PR do?**
Create new recipe.

**Description of Task to be completed?**
- write unit test
- defined the recipe schema
- implement create article

**How should this be manually tested?**
- clone this repository
- run `npm install`
- start the dev server, 'npm run devstart`
- open up postman and generate an existing user's token using the endpoint, 'localhost:{PORT}/user/login. Ensure the email and password of the user is provided.
- add the token to the req.headers with 'authorization' as the key and the endpoint to 'localhost:{PORT}/recipe

**Any background context you want to provide?**
- New recipes can only be created by a registered user

**What are the relevant pivotal tracker stories?**
#164139774

**Screenshots (if appropriate)**

[#164139774]

Commit:
feat(create-recipe): create recipe
- write unit test
- defined the recipe schema
- implement create article
[#164139774]

andela-pessien commented on server/controllers/recipe-controller.js:
```diff
@@ -0,0 +1,68 @@
+      steps,
+      cookingTime,
+      preparationTime,
+      tagList
```
This isn't your task :)

blackdevelopa replied:
Okay. Noted. I have fixed this.

andela-pessien commented on server/migrations/20190312121718-create-recipe.js:
```diff
@@ -0,0 +1,46 @@
+      slug: {
+        type: Sequelize.STRING
+      },
+      ingredients: {
```
**[question]** Will we be storing the ingredients and their amounts together?

blackdevelopa replied:
Yes. We will store the ingredients and the amounts together. An example will be, `1/2 cups of rice`. The ingredients will be an array of strings.

andela-pessien replied:
👍 We'll be further discussing this schema though as more features get built out

andela-pessien commented on server/utils/initdb.js:
```diff
@@ -1,5 +1,8 @@
+import envConfig from '../config/env-config';
 
-db.sequelize.sync().then(() => {
+const params = envConfig.NODE_ENV === 'test' ? { force: true } : {};
```
**[comment]** I don't expect us to keep syncing tables in production.

blackdevelopa replied:
We are working on this. This will be fixed in another PR.

andela-pessien commented on test/recipes.spec.js:
```diff
@@ -0,0 +1,167 @@
+          expect(res).to.have.status(201);
+          expect(body.recipe).to.be.an('object');
+          expect(body.recipe.id).to.be.a('number');
+          expect(body.recipe.title).to.be.a('string');
```
Shouldn't we test that the attributes returned match what was sent?

blackdevelopa replied:
Okay. I will test that the response is same as the input.

andela-pessien commented on test/recipes.spec.js:
```diff
@@ -0,0 +1,167 @@
+        .end((err, res) => {
+          const { body } = res;
+          expect(res).to.have.status(404);
+          expect(body.errors).to.contain('Title must be present');
```
The error should match the offense :)

blackdevelopa replied:
Yes. The code that solved for this was lost after my first rebase. This is fixed now.

Commit:
feat(createRecipe): Create Recipe
dquote> - add new unit test
dquote> - integrate with login feature
dquote> [#164139774]

Commit:
feat(createRecipe): Create Recipe
dquote> - modified unit test
dquote> [Finishes #164139774]

Commit:
poking at travis

Commit:
poking at travis some more

Commit:
poking at travis some more

Commit:
poking at travis some more

andela-pessien commented on server/migrations/20190312121718-create-recipe.js:
```diff
@@ -0,0 +1,42 @@
+      preparationTime: {
+        type: Sequelize.INTEGER
+      },
+      videoList: {
```
We don't yet seem to be using this, plus wouldn't videos have the same challenges we highlighted with images?

blackdevelopa replied:
Not exactly, we will be using an array of strings instead. The intention is that the videos will belong to the recipe and not the steps.

andela-pessien commented on server/models/recipe.js:
```diff
@@ -0,0 +1,25 @@
+      steps: DataTypes.JSONB,
+      cookingTime: DataTypes.INTEGER,
+      preparationTime: DataTypes.INTEGER,
+      videoList: DataTypes.STRING
```
Related to the above, would've thought this would be an array at least

blackdevelopa replied:
Yes. VideoList will be an array of strings. I have implemented this.

andela-pessien commented on server/models/recipe.js:
```diff
@@ -0,0 +1,25 @@
+    },
+    {}
+  );
+  Recipe.associate = () => {
```
How come the association between recipes and users isn't defined here?

blackdevelopa replied:
Okay. This is valid. I will implement this right away

andela-pessien commented on server/utils/recipes.js:
```diff
@@ -0,0 +1,26 @@
+    .required(),
+  steps: Joi.object()
+    .keys({
+      1: Joi.object().keys({
```
Only validates key `1`?

andela-pessien commented on server/utils/slugify.js:
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable no-useless-escape */
+const slugifyTitle = string => {
+  const a = 'àáäâãåăæçèéëêǵḧìíïîḿńǹñòóöôœṕŕßśșțùúüûǘẃẍÿź·/_,:;';
```
In this house we do not host single-letter variable names (except _maybe_ in for loops 🙂)

You should still be able to comfortably explain this code if we come back to it at the end of Simulations

blackdevelopa replied:
I have updated this function.

Commit:
Merge branch 'ft-create-recipe-164139774' of https://github.com/andela/vale-ah-backend into ft-create-recipe-164139774

andela-pessien reviewed (approved):
Looks good to me but please don't forget to reach a consensus concerning what happens to recipes of deleted users
